### PR TITLE
prepared queries

### PR DESCRIPTION
**Pull request by prithvi081099**

### CODE REVIEW

File: gateway/modules/crud/operations.go [hunk 1]
```diff
@@ -160,6 +160,32 @@ func (m *Module) Delete(ctx context.Context, dbAlias, project, col string, req *
...
+	if err != nil {
+		return err
+	}
+
+	if err := crud.IsClientSafe(); err != nil {
+		return err
+	}
+
+	var args []interface{}
+	for i := 0; i < len(m.queries.Arguments); i++ {
```
Comment (YourTechBud) [inline gateway/modules/crud/operations.go:178]: m.queries will be a map. You'll have to first check if the provided id exists in the map. Throw error if it doesn't

File: gateway/modules/crud/sql/read.go [hunk 2]
```diff
@@ -123,7 +123,11 @@ func (s *SQL) read(ctx context.Context, project, col string, req *model.ReadRequ
 
 	logrus.Debugf("Executing sql read query: %s - %v", sqlString, args)
 
-	stmt, err := executor.PreparexContext(ctx, sqlString)
+	return s.readexec(ctx, sqlString, args, req.Operation, executor)
+}
+
+func (s *SQL) readexec(ctx context.Context, project string, args []interface{}, req string, executor executor) (int64, interface{}, error) {
```
Comment (YourTechBud) [inline gateway/modules/crud/sql/read.go:129]: Don't call the variable req. Call it operation

File: gateway/utils/server/routes.go [hunk 3]
```diff
@@ -56,9 +56,12 @@ func (s *Server) routes(profiler bool, staticPath string, restrictedHosts []stri
 	router.Methods(http.MethodGet).Path("/v1/config/projects/{project}/database/collections/rules").HandlerFunc(handlers.HandleGetTableRules(s.adminMan, s.syncMan))
 	router.Methods(http.MethodGet).Path("/v1/config/projects/{project}/database/config").HandlerFunc(handlers.HandleGetDatabaseConfig(s.adminMan, s.syncMan))
 	router.Methods(http.MethodGet).Path("/v1/config/projects/{project}/database/collections/schema/mutate").HandlerFunc(handlers.HandleGetSchemas(s.adminMan, s.syncMan))
+	router.Methods(http.MethodGet).Path("/v1/config/projects/{project}/database/prepared-queries").HandlerFunc(handlers.HandleGetPreparedQuery(s.adminMan, s.syncMan))
```
Comment (YourTechBud) [inline gateway/utils/server/routes.go:59]: Keep all the routes you create together

File: gateway/utils/server/routes.go [hunk 4]
```diff
@@ -56,9 +56,12 @@ func (s *Server) routes(profiler bool, staticPath string, restrictedHosts []stri
...
 	router.Methods(http.MethodPost).Path("/v1/config/projects/{project}/database/{dbAlias}/collections/{col}/rules").HandlerFunc(handlers.HandleSetTableRules(s.adminMan, s.syncMan))
 	router.Methods(http.MethodPost).Path("/v1/config/projects/{project}/database/{dbAlias}/config/{id}").HandlerFunc(handlers.HandleSetDatabaseConfig(s.adminMan, s.syncMan))
 	router.Methods(http.MethodDelete).Path("/v1/config/projects/{project}/database/{dbAlias}/config/{id}").HandlerFunc(handlers.HandleRemoveDatabaseConfig(s.adminMan, s.syncMan))
+	router.Methods(http.MethodPost).Path("/v1/config/projects/{project}/database/{dbAlias}/config/{id}").HandlerFunc(handlers.HandleSetPreparedQueries(s.adminMan, s.syncMan))
```
Comment (YourTechBud) [inline gateway/utils/server/routes.go:63]: I won't accept such silly mistakes next time
```suggestion
	router.Methods(http.MethodPost).Path("/v1/config/projects/{project}/database/{dbAlias}/prepared-queries/{id}").HandlerFunc(handlers.HandleSetPreparedQueries(s.adminMan, s.syncMan))
```

File: gateway/utils/syncman/syncman_crud.go [hunk 5]
```diff
@@ -65,6 +65,35 @@ func (s *Manager) SetDatabaseConnection(ctx context.Context, project, dbAlias st
 	return s.setProject(ctx, projectConfig)
 }
 
+// SetPreparedQueries sets database preparedqueries
+func (s *Manager) SetPreparedQueries(ctx context.Context, project, dbAlias, id string, v config.PreparedQuery) error {
```
Comment (YourTechBud) [inline gateway/utils/syncman/syncman_crud.go:69]: Keep all your functions together

File: gateway/utils/syncman/syncman_crud.go [hunk 6]
```diff
@@ -65,6 +65,35 @@ func (s *Manager) SetDatabaseConnection(ctx context.Context, project, dbAlias st
...
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	projectConfig, err := s.getConfigWithoutLock(project)
+	if err != nil {
+		return err
+	}
+
+	// update database config
+	coll, ok := projectConfig.Modules.Crud[dbAlias].PreparedQueries[id]
```
Comment (YourTechBud) [inline gateway/utils/syncman/syncman_crud.go:80]: Incorrect logic

File: gateway/utils/graphql/graphql.go [hunk 7]
```diff
@@ -173,6 +173,21 @@ func (graph *Module) execGraphQLDocument(ctx context.Context, node ast.Node, tok
...
 
+			if kind == "prepared-queries" {
+				graph.execPreparedQueryRequest(ctx, field, token, store, createDBCallback(func(dbAlias, col string, result interface{}, err error) {
+					if err != nil {
+						cb(nil, err)
+						return
+					}
+
+					// Load the schema
+					s, _ := graph.schema.GetSchema(dbAlias, col)
```
Comment (YourTechBud) [inline gateway/utils/graphql/graphql.go:184]: Not needed. Remove this

File: gateway/utils/graphql/graphql.go [hunk 8]
```diff
@@ -225,9 +240,12 @@ func (graph *Module) execGraphQLDocument(ctx context.Context, node ast.Node, tok
 	}
 }
 
-func (graph *Module) getQueryKind(directive string) string {
+func (graph *Module) getQueryKind(directive, id string) string {
```
Comment (YourTechBud) [inline gateway/utils/graphql/graphql.go:243]: Rename `id` to fieldName

File: gateway/utils/graphql/read.go [hunk 9]
```diff
@@ -67,6 +67,38 @@ func (graph *Module) execReadRequest(ctx context.Context, field *ast.Field, toke
...
 }
 
+func (graph *Module) execPreparedQueryRequest(ctx context.Context, field *ast.Field, token string, store utils.M, cb dbCallback) {
+	dbAlias, err := graph.GetDBAlias(field)
+	if err != nil {
+		cb("", "", nil, err)
+		return
+	}
+
+	id, err := getCollection(field)
```
Comment (YourTechBud) [inline gateway/utils/graphql/read.go:77]: Id is nothing but `field.Name.Value`

File: gateway/utils/graphql/read.go [hunk 10]
```diff
@@ -67,6 +67,38 @@ func (graph *Module) execReadRequest(ctx context.Context, field *ast.Field, toke
...
+	if err != nil {
+		cb("", "", nil, err)
+		return
+	}
+	param, err := getFuncParams(field, store)
```
Comment (YourTechBud) [inline gateway/utils/graphql/read.go:82]: Rename param to params

File: gateway/utils/handlers/config_crud.go [hunk 11]
```diff
@@ -256,6 +256,123 @@ func HandleRemoveDatabaseConfig(adminMan *admin.Manager, syncman *syncman.Manage
...
+		dbAliasQuery, exists := r.URL.Query()["dbAlias"]
+		if exists {
+			dbAlias = dbAliasQuery[0]
+		}
+		idQuery, exists := r.URL.Query()["id"]
+		id := ""
+		if exists {
+			id = idQuery[0]
+		}
+		schemas, err := syncMan.GetPreparedQuery(ctx, projectID, dbAlias, id)
```
Comment (YourTechBud) [inline gateway/utils/handlers/config_crud.go:287]: Rename variable schemas to result

File: gateway/utils/handlers/crud.go [hunk 12]
```diff
@@ -18,6 +18,57 @@ type requestMetaData struct {
...
+		// Create a context of execution
+		ctx, cancel := context.WithTimeout(r.Context(), 30*time.Second)
+		defer cancel()
+
+		// Get the path parameters
+		vars := mux.Vars(r)
+		dbAlias := vars["dbAlias"]
+		project := vars["project"]
+		id := vars["id"]
+		token := vars["token"]
```
Comment (YourTechBud) [inline gateway/utils/handlers/crud.go:37]: You get the token from the headers. Use getTokenFromRequest method.

File: gateway/utils/syncman/syncman_crud.go [hunk 13]
```diff
@@ -87,6 +87,106 @@ func (s *Manager) RemoveDatabaseConfig(ctx context.Context, project, dbAlias str
...
+	}
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	projectConfig, err := s.getConfigWithoutLock(project)
+	if err != nil {
+		return nil, err
+	}
+	if dbAlias != "" && id != "" {
+		preparedQueries, ok := projectConfig.Modules.Crud[dbAlias].PreparedQueries[id]
```
Comment (YourTechBud) [inline gateway/utils/syncman/syncman_crud.go:104]: Check if dbAlias is present or not

File: gateway/utils/syncman/syncman_crud.go [hunk 14]
```diff
@@ -87,6 +87,106 @@ func (s *Manager) RemoveDatabaseConfig(ctx context.Context, project, dbAlias str
...
+		if !ok {
+			return nil, fmt.Errorf("collection (%s) not present in config for dbAlias (%s) )", dbAlias, id)
```
Comment (YourTechBud) [inline gateway/utils/syncman/syncman_crud.go:106]: Are you even reading the error messages?

File: gateway/utils/syncman/syncman_crud.go [hunk 15]
```diff
@@ -87,6 +87,106 @@ func (s *Manager) RemoveDatabaseConfig(ctx context.Context, project, dbAlias str
 	return s.setProject(ctx, projectConfig)
 }
 
+// GetPreparedQuery gets preparedQuery from config
+func (s *Manager) GetPreparedQuery(ctx context.Context, project, dbAlias, id string) ([]interface{}, error) {
+	// Acquire a lock
+	type response struct {
+		SQL string `json:"sql"`
```
Comment (YourTechBud) [inline gateway/utils/syncman/syncman_crud.go:94]: This should contain id and arguments as well

File: gateway/utils/syncman/syncman_crud.go [hunk 16]
```diff
@@ -87,6 +87,106 @@ func (s *Manager) RemoveDatabaseConfig(ctx context.Context, project, dbAlias str
...
+	projectConfig, err := s.getConfigWithoutLock(project)
+	if err != nil {
+		return nil, err
+	}
+	if dbAlias != "" && id != "" {
+		preparedQueries, ok := projectConfig.Modules.Crud[dbAlias].PreparedQueries[id]
+		if !ok {
+			return nil, fmt.Errorf("collection (%s) not present in config for dbAlias (%s) )", dbAlias, id)
+		}
+		return []interface{}{map[string]*response{fmt.Sprintf("%s-%s", dbAlias, id): {SQL: preparedQueries.SQL}}}, nil
```
Comment (YourTechBud) [inline gateway/utils/syncman/syncman_crud.go:108]: Return response. Why are you making a map?

File: gateway/utils/syncman/syncman_crud.go [hunk 17]
```diff
@@ -87,6 +87,106 @@ func (s *Manager) RemoveDatabaseConfig(ctx context.Context, project, dbAlias str
...
+	} else if dbAlias != "" {
+		preparedQueries := projectConfig.Modules.Crud[dbAlias].PreparedQueries
```
Comment (YourTechBud) [inline gateway/utils/syncman/syncman_crud.go:110]: Logic is incorrect

File: gateway/utils/syncman/syncman_crud.go [hunk 18]
```diff
@@ -87,6 +87,106 @@ func (s *Manager) RemoveDatabaseConfig(ctx context.Context, project, dbAlias str
...
+		coll := map[string]*response{}
+		for key, value := range preparedQueries {
+			coll[fmt.Sprintf("%s-%s", dbAlias, key)] = &response{SQL: value.SQL}
+		}
+		return []interface{}{coll}, nil
+	}
+	databases := projectConfig.Modules.Crud
+	coll := map[string]*response{}
```
Comment (YourTechBud) [inline gateway/utils/syncman/syncman_crud.go:118]: Logic is incorrect

File: gateway/utils/syncman/syncman_crud.go [hunk 19]
```diff
@@ -87,6 +87,106 @@ func (s *Manager) RemoveDatabaseConfig(ctx context.Context, project, dbAlias str
...
+	if err != nil {
+		return err
+	}
+
+	// update database PreparedQueries
+	databaseConfig, ok := projectConfig.Modules.Crud[dbAlias]
+	if !ok {
+		return errors.New("specified database not present in config")
+	}
+	// preparedQuery, ok := databaseConfig.PreparedQueries[id]
```
Comment (YourTechBud) [inline gateway/utils/syncman/syncman_crud.go:143]: Remove the comments

File: gateway/utils/syncman/syncman_crud.go [hunk 20]
```diff
@@ -87,6 +87,106 @@ func (s *Manager) RemoveDatabaseConfig(ctx context.Context, project, dbAlias str
...
+
+	projectConfig, err := s.getConfigWithoutLock(project)
+	if err != nil {
+		return err
+	}
+
+	// update database PreparedQueries
+	databaseConfig, ok := projectConfig.Modules.Crud[dbAlias]
+	if !ok {
+		return errors.New("specified database not present in config")
```
Comment (YourTechBud) [inline gateway/utils/syncman/syncman_crud.go:141]: Include db alias in error message

File: gateway/utils/syncman/syncman_crud.go [hunk 21]
```diff
@@ -87,6 +87,106 @@ func (s *Manager) RemoveDatabaseConfig(ctx context.Context, project, dbAlias str
...
+	// 	}
+	// } else {
+	// 	preparedQuery.ID = v.ID
+	// 	preparedQuery.SQL = v.SQL
+	// 	preparedQuery.Arguments = v.Arguments
+	// }
+	if databaseConfig.PreparedQueries == nil {
+		databaseConfig.PreparedQueries = make(map[string]*config.PreparedQuery, 1)
+	}
+	databaseConfig.PreparedQueries = map[string]*config.PreparedQuery{id: v}
```
Comment (YourTechBud) [inline gateway/utils/syncman/syncman_crud.go:158]: Why are you resetting the map? What are you doing?

File: gateway/utils/syncman/syncman_crud.go [hunk 22]
```diff
@@ -87,6 +87,106 @@ func (s *Manager) RemoveDatabaseConfig(ctx context.Context, project, dbAlias str
...
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	projectConfig, err := s.getConfigWithoutLock(project)
+	if err != nil {
+		return err
+	}
+
+	// update database reparedQueries
+	delete(projectConfig.Modules.Crud[dbAlias].PreparedQueries, id)
```
Comment (YourTechBud) [inline gateway/utils/syncman/syncman_crud.go:180]: Check if dbAlias is present

File: gateway/modules/crud/operations.go [hunk 23]
```diff
@@ -160,6 +160,42 @@ func (m *Module) Delete(ctx context.Context, dbAlias, project, col string, req *
 	return err
 }
 
+// IsPreparedQueryPresent checks if id exist
+func (m *Module) IsPreparedQueryPresent(id string) bool {
+	_, p := m.queries[id]
```
Comment (YourTechBud) [inline gateway/modules/crud/operations.go:165]: Acquire a RLock before doing this

File: gateway/modules/crud/operations.go [hunk 24]
```diff
@@ -160,6 +160,42 @@ func (m *Module) Delete(ctx context.Context, dbAlias, project, col string, req *
...
+	}
+
+	if err := crud.IsClientSafe(); err != nil {
+		return nil, err
+	}
+
+	var args []interface{}
+	preparedQuery, p := m.queries[id]
+	if !p {
+		return nil, fmt.Errorf("Prepared Query for given id:{%s} does not exist", id)
```
Comment (YourTechBud) [inline gateway/modules/crud/operations.go:186]:
```suggestion
		return nil, fmt.Errorf("Prepared Query for given id (%s) does not exist", id)
```

File: gateway/modules/crud/crud.go [hunk 25]
```diff
@@ -28,7 +28,7 @@ type Module struct {
 	project            string
 	removeProjectScope bool
 	schema             model.SchemaCrudInterface
-
+	queries            map[string]*config.PreparedQuery
```
Comment (YourTechBud) [inline gateway/modules/crud/crud.go:31]: When do you set this?

File: gateway/utils/handlers/config_crud.go [hunk 26]
```diff
@@ -256,6 +256,123 @@ func HandleRemoveDatabaseConfig(adminMan *admin.Manager, syncman *syncman.Manage
...
+			id = idQuery[0]
+		}
+		result, err := syncMan.GetPreparedQuery(ctx, projectID, dbAlias, id)
+		if err != nil {
+			w.Header().Set("Content-Type", "application/json")
+			w.WriteHeader(http.StatusInternalServerError)
+			_ = json.NewEncoder(w).Encode(map[string]string{"error": err.Error()})
+			return
+		}
+		w.WriteHeader(http.StatusOK)
```
Comment (YourTechBud) [inline gateway/utils/handlers/config_crud.go:294]: Set the json content type header

File: gateway/utils/handlers/config_crud.go [hunk 27]
```diff
@@ -256,6 +256,123 @@ func HandleRemoveDatabaseConfig(adminMan *admin.Manager, syncman *syncman.Manage
...
+			w.Header().Set("Content-Type", "application/json")
+			w.WriteHeader(http.StatusInternalServerError)
+			_ = json.NewEncoder(w).Encode(map[string]string{"error": err.Error()})
+			return
+		}
+
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusOK) // http status codee
+		_ = json.NewEncoder(w).Encode(map[string]interface{}{})
+		// return
```
Comment (YourTechBud) [inline gateway/utils/handlers/config_crud.go:372]: Remove this comment

File: gateway/utils/syncman/syncman_crud.go [hunk 28]
```diff
@@ -87,6 +87,109 @@ func (s *Manager) RemoveDatabaseConfig(ctx context.Context, project, dbAlias str
...
+		if !ok {
+			return nil, fmt.Errorf("specified database:{%s} not present in config", dbAlias)
+		}
+
+		if id != "" {
+			preparedQuery, ok := databaseConfig.PreparedQueries[id]
+			if !ok {
+				return nil, fmt.Errorf("Prepared Queries for id (%s) not present in config for dbAlias (%s) )", id, dbAlias)
+			}
+			return []interface{}{&response{ID: preparedQuery.ID, SQL: preparedQuery.SQL, Arguments: preparedQuery.Arguments}}, nil
```
Comment (YourTechBud) [inline gateway/utils/syncman/syncman_crud.go:117]: Same mistake even after we spoke about this. I'm disappointed
```suggestion
			return []interface{}{&response{ID: id, SQL: preparedQuery.SQL, Arguments: preparedQuery.Arguments}}, nil
```

File: gateway/utils/syncman/syncman_crud.go [hunk 29]
```diff
@@ -87,6 +87,109 @@ func (s *Manager) RemoveDatabaseConfig(ctx context.Context, project, dbAlias str
...
+		}
+		preparedQuery := databaseConfig.PreparedQueries
+		var coll []interface{} = make([]interface{}, 0)
+		for key, value := range preparedQuery {
+			coll = append(coll, &response{ID: key, SQL: value.SQL, Arguments: value.Arguments})
+		}
+		return []interface{}{coll}, nil
```
Comment (YourTechBud) [inline gateway/utils/syncman/syncman_crud.go:124]: Did you even once try consdering why we always return an array with a single element?
```suggestion
		return coll, nil
```

File: gateway/utils/syncman/syncman_crud.go [hunk 30]
```diff
@@ -87,6 +87,109 @@ func (s *Manager) RemoveDatabaseConfig(ctx context.Context, project, dbAlias str
...
+	}
+	databases := projectConfig.Modules.Crud
+	var coll []interface{} = make([]interface{}, 0)
+	for _, dbInfo := range databases {
+		for key, value := range dbInfo.PreparedQueries {
+			coll = append(coll, &response{ID: key, SQL: value.SQL, Arguments: value.Arguments})
+		}
+	}
+	return []interface{}{coll}, nil
```
Comment (YourTechBud) [inline gateway/utils/syncman/syncman_crud.go:133]:
```suggestion
	return coll, nil
```

**Comment by sharadregoti** [comment]
@YourTechBud i have tested prepared queries

**Comment by YourTechBud** [comment]
Fixes #815

**Pull request merged**